### ISSUE TITLE
`<xutility>`: `default_initializable<move_iterator<It>>` may be problematic for Clang

### ISSUE THREAD

**Issue by frederick-vs-ja** [microsoft/STL#3377]
**Describe the bug**
TL;DR: LLVM-60293 is likely not a bug and we may need to change `move_iterator`'s default constructor to make Clang happy.

In cases like `views::join_with(r1 | views::as_rvalue, r2 | views::as_rvalue)` (already occur in [test file](https://github.com/microsoft/STL/blob/f8b8ad6dea901d360d8165cc6719de47ca9d9ca2/tests/std/tests/P2441R2_views_join_with/test.cpp) due to #3359), the iterator type of the resulted `join_with_view` wraps an `move_iterator` and `default_initializable<move_iterator<It>>` is detected, where `It` is the iterator type of `r1`. When `It` is a non-default-initializable `input_iterator` type (more precisely, when `It{}` is ill-formed), MSVC STL's implementation of `move_iterator`results in weird behavior of Clang.

MSVC STL's `move_iterator` has a defaulted default constructor:
https://github.com/microsoft/STL/blob/f8b8ad6dea901d360d8165cc6719de47ca9d9ca2/stl/inc/xutility#L3942
And the required value-initialization is performed via default member initializer:
https://github.com/microsoft/STL/blob/f8b8ad6dea901d360d8165cc6719de47ca9d9ca2/stl/inc/xutility#L4165

Per [[class.default.ctor]/2.3](https://eel.is/c++draft/class.default.ctor#2.3), the default member initializer makes the default constructor of `move_iterator` never deleted (which is decided in [CWG-927](https://cplusplus.github.io/CWG/issues/927.html)). Given a non-default-initializable `It`, `default_initializable<move_iterator<It>>` ask whether `is_constructible_v<move_iterator<It>>` is `true`, which may make the program ill-formed per [this note](https://eel.is/c++draft/meta.rel#note-3), and implementation divergence seems (unfortunately) permitted (discussed in llvm/llvm-project#60293).

There're many iterator-related facilities using `default_initializable` to constrain their default constructors. We may need to change `move_iterator`'s default constructor to make them usable with `move_iterator` and non-default-initializable input iterators when compiling with Clang.

The current approach has some benefits:
- the exception specification is automatically strengthened;
- the default constructor can be sometimes implicitly `constexpr` in C++14 mode, although this makes little sense because most operations are only `constexpr` since C++17.

Possible resolutions (may be only applied for Clang):
- Just making the default constructor non-defaulted. Cons:
  - either we need a non-conforming `constexpr` in C++14 mode, or we'll break some non-portable `constexpr` uses of `move_iterator` in C++14;
  - exception specification strengthening may be preferable, but throughput might be (sightly) damaged.
- Adding base classes to make the default constructor conditionally deleted and otherwise defaulted. Cons:
  - also damages throughput;
  - the adapted iterator may be designed in such way that asking whether it satisfies `is_default_constructible` is ill-formed, while neither `input_iterator` nor _Cpp17InputIterator_ cares default-initializability.

**Comment by StephanTLavavej** [comment]
@CaseyCarter notes that many occurrences are affected, not just `move_iterator`. We talked about this at the weekly maintainer meeting and we want to wait until the Clang issue has a definitive resolution (either fixed or by design) before proceeding here. (We probably don't need to use temporary workarounds until the Clang fix ships, because this hasn't been encountered during normal usage, only during development of `<ranges>`.)  Personally, I am willing to accept throughput costs here, so I don't think that should affect whatever course of action we choose.

**Comment by frederick-vs-ja** [comment]
LWG-4125 touches this.
